skcosmo/feature_selection/voronoi_fps.py

```diff
@@ -0,0 +1,163 @@
+import numpy as np
+
+from .simple_fps import SimpleFPS
+
+VORONOI_CUTOFF_FRACTION = 1.0 / 6.0
+
+
+class VoronoiFPS(SimpleFPS):
+    """
+    Base Class defined for Voronoi FPS methods
+
+    :param initialize: predetermined index; if None provided, first index selected
+                 is 0
+    :type selected_: int, None
+    """
+
+    def _init_greedy_search(self, X, y, n_to_select):
+
+        n_features = X.shape[1]
+
+        # voronoi polyhedron location of each index
+        self.vlocation_of_idx = np.full(n_features, -1)
+
+        # indices in each voronoi polyhedron
+        self.idx_in_voronoi = [np.array([]) for n in range(self.n_features_to_select)]
+        self.idx_in_voronoi[0] = np.arange(n_features, dtype=int)
+
+        # number of points in each voronoi polyhedron (VP)
+        self.number_in_voronoi = np.full(self.n_features_to_select, 0)
+        self.number_in_voronoi[0] = n_features
+
+        # furthest point in from each VP center
+        self.furthest_point = np.full(self.n_features_to_select, -1, dtype=int)
+
+        # quarter of the square distance between new selected point and previously
+        # selected points
+        self.sel_d2q_ = np.zeros(self.n_features_to_select, float)
+
+        super()._init_greedy_search(X, y, n_to_select)
+
+    def _continue_greedy_search(self, X, y, n_to_select):
+        """ Continues the search. Prepares an array to store the selected features. """
+
+        super()._continue_greedy_search(X, y, n_to_select)
+
+        n_pad = n_to_select - self.n_selected_
+        for n in range(n_pad):
+            self.idx_in_voronoi.append(np.array([]))
+        self.number_in_voronoi = np.pad(
+            self.number_in_voronoi, (0, n_pad), "constant", constant_values=0
+        )
+        self.sel_d2q_ = np.pad(self.sel_d2q_, (0, n_pad), "constant", constant_values=0)
+        self.furthest_point = np.concatenate(
+            (self.furthest_point, np.zeros(n_pad, dtype=int))
+        )
+
+    def _get_active(self, X, last_selected):
+        # because we haven't called super() yet, self.n_selected_ has not been
+        # incremented yet
+
+        f_active = np.zeros(self.n_features_to_select, bool)
+
+        """must compute distance of the new point to all the previous FPS. Some
+            of these might have been computed already, but bookkeeping could be
+            worse that recomputing (TODO: verify!)
+            """
+        # calculation in a single block
+        self.sel_d2q_[: self.n_selected_] = (
+            self.norms_[self.selected_idx_[: self.n_selected_]]
+            + self.norms_[last_selected]
+            - 2 * (self.X_selected_[:, : self.n_selected_].T @ X[:, last_selected])
+        ) * 0.25
+
+        for ic in range(self.n_selected_):
+            # empty voronoi, no need to consider it
+            if self.number_in_voronoi[ic] > 1:
+                r2 = max(self.haussdorf_[self.idx_in_voronoi[ic]])
```

Comment:
The same issue as last time remains here, plus an addition. We've already introduced the self.furthest_point array, we handle it everywhere, but I'm not quite sure why we need it. How it's used, and why it's not used in a place where it could be useful.